clients/pkg/promtail/client/watcher.go

```diff
@@ -0,0 +1,304 @@
+package client
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/tsdb/record"
+	"github.com/prometheus/prometheus/tsdb/wlog"
+
+	"github.com/grafana/loki/pkg/ingester"
+)
+
+const (
+	readPeriod         = 10 * time.Millisecond
+	checkpointPeriod   = 5 * time.Second
+	segmentCheckPeriod = 100 * time.Millisecond
+)
```

Comment:
Something to be aware of here:

I've found that this adds a 4-5% CPU usage to the baseline resource consumption of the process. See prometheus/prometheus#11625.

The CPU utilization is negligible if something is writing to the WAL frequently enough, but if writes to the WAL are infrequent (i.e., slow moving log files), then the CPU usage will be higher than expected.

This could impact users with limited hardware capabilities, like Raspberry Pis.